x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java

```diff
@@ -236,6 +238,19 @@ private static Collection<Failure> validateRow(Row row) {
         return failures;
     }
 
+    private static Collection<Failure> validateEval(Eval eval) {
+        List<Failure> failures = new ArrayList<>(eval.fields().size());
+        eval.fields().forEach(field -> {
+            DataType dataType = field.dataType();
+            if (EsqlDataTypes.isRepresentable(dataType) == false) {
+                failures.add(
+                    fail(field, "EVAL does not support type [{}] in expression [{}]", dataType.typeName(), field.child().sourceText())
```

Comment:
Wondering if for numeric types we shouldn't suggest a conversion, à la: `cannot use [{}]  of type [{}] in an EVAL assignment. Convert to a wider type`.

Reply:
Hm, this would only apply to some of the [unrepresentable data types](https://github.com/elastic/elasticsearch/blob/c663c831b7e334f980667686d1b0555a605f4bdf/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java#L151). Converting to a wider type does not really apply to a date period, for instance, right?

But, we could add a couple more type specific error messages, like "convert to a wider type" for short, half float etc. and "can use period/duration only in combination with a datetime" or so for periods/durations.

I planned to do another small PR once this is merged to have good error messages for cases when users add durations and periods (which will inevitably happen IMO until we implement mixed durations and periods). I can add type-specific verifier errors for EVAL there :)

Reply:
> Hm, this would only apply to some of the [unrepresentable data types](https://github.com/elastic/elasticsearch/blob/c663c831b7e334f980667686d1b0555a605f4bdf/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java#L151). Converting to a wider type does not really apply to a date period, for instance, right?

Right, just for numeric types (`DataType#isNumeric()`).

> when users add durations and periods

Right, we can suggest reordering there.